<a href="https://colab.research.google.com/github/Ryan-M-Smith/CS315/blob/main/InClass/quicksort.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
# Template for testing Quick-Sort on small lists, outputing each intermediate step

In [2]:
import random
import time
import numpy as np
from typing import Any

In [3]:
# Partition provided as a convenience

def partition(arr: list[Any], p: int, r: int) -> int:
    x = arr[r]
    i = p - 1

    for j in range(p,r):
        if arr[j] <= x:
            i += 1
            arr[i], arr[j] = arr[j], arr[i]    # Swap arr[i] and arr[j]

    arr[i+1], arr[r] = arr[r], arr[i+1]        # Swap arr[i+1] and arr[r]

    return i + 1

In [4]:
def quicksort(arr: list[Any], p: int, r: int, *, print_flag: bool = False) -> None:
    if p < r:
        q = partition(arr, p, r)

        if print_flag:
            print(arr)

        quicksort(arr, p, q - 1, print_flag=print_flag)
        quicksort(arr, q + 1 , r, print_flag=print_flag)

In [5]:
n = 8
arr = list(range(n))
random.shuffle(arr)

print(arr)
quicksort(arr, 0, n - 1, print_flag=True)
print(arr)

[6, 0, 4, 1, 7, 5, 2, 3]
[0, 1, 2, 3, 7, 5, 4, 6]
[0, 1, 2, 3, 7, 5, 4, 6]
[0, 1, 2, 3, 7, 5, 4, 6]
[0, 1, 2, 3, 5, 4, 6, 7]
[0, 1, 2, 3, 4, 5, 6, 7]
[0, 1, 2, 3, 4, 5, 6, 7]


In [ ]:
#
# HW04 - time Quicksort on random data for a variety of array sizes.
#

rng = np.random.default_rng()
TEST_RUNS = 10
sizes = [10, 50, 100, 1000, 2500, 7500, 10000, 12500, 17500, 20000]

for i in range(TEST_RUNS):
        n = sizes[i]
        arr = rng.integers(low=0, high=100, size=n).tolist()

        start = time.perf_counter()
        quicksort(arr, 0, n - 1)
        end = time.perf_counter()

        print(f"n = {n:<8,} time = {(end - start) * 1_000:.3f} ms")

n = 10       time = 0.006 ms
n = 50       time = 0.022 ms
n = 100      time = 0.900 ms
n = 1,000    time = 1.816 ms
n = 2,500    time = 5.161 ms
n = 7,500    time = 26.336 ms
n = 10,000   time = 45.511 ms
n = 12,500   time = 63.852 ms
n = 17,500   time = 118.799 ms
n = 20,000   time = 153.088 ms


## Results

Because Quicksort Sort has an expected runtime of $\Theta\left(n \lg n\right)$, we will calculate all ratios with a denominator of $n \lg n$.

| $n$    | $T(n)$      | $\dfrac{T(n)}{n \lg n}$      |
| ------ | ----------- | ---------------------------- |
| 10     | 0.006       | 0.000180                     |
| 50     | 0.022       | 0.000538                     |
| 100    | 0.900       | 0.01354                      |
| 1,000  | 1.816       | 0.000182                     |
| 2,500  | 5.161       | 0.000183                     |
| 7,500  | 26.336      | 0.000272                     |
| 10,000 | 45.511      | 0.000271                     |
| 12,500 | 63.852      | 0.000374                     |
| 17,500 | 118.799     | 0.000540                     |
| 20,000 | 153.088     | 0.000464                     |



Looking at our ratios, as $n$ gets larger, the results tend to stabilize around a factor of $1.0 \times 10^{-4}$, showing that the runtime tends to stabilize
around a ratio of $n \lg n$. This strongly suggests that the algorithm runs in $\Theta\left(n \lg n\right)$ time.


In [ ]:
#
# HW04 - time Quicksort on sorted data for a variety of array sizes.
#

# This code will run into recursion depth issues for large n because quicksort
# will be running in a worst-case scenario.
import sys
sys.setrecursionlimit(25000)

for i in range(TEST_RUNS):
    n = sizes[i]
    arr = list(range(n))

    start = time.perf_counter()
    quicksort(arr, 0, n - 1)
    end = time.perf_counter()

    print(f"n = {n:<8,} time = {(end - start)*1_000:.3f} ms")

n = 10       time = 0.008 ms
n = 50       time = 0.079 ms
n = 100      time = 0.302 ms
n = 1,000    time = 30.955 ms
n = 2,500    time = 201.886 ms
n = 7,500    time = 1923.617 ms
n = 10,000   time = 3299.818 ms
n = 12,500   time = 5247.543 ms
n = 17,500   time = 9953.997 ms
n = 20,000   time = 13118.797 ms


In [8]:
def quicksort_no_tr(arr: list[Any], p: int, r: int, *, print_flag: bool = False) -> None:
    while p < r:
      q = partition(arr, p, r)

      if print_flag:
        print(arr)

      quicksort(arr, p, q - 1, print_flag=print_flag)
      p = q + 1

n = 8
arr = list(range(n))
random.shuffle(arr)

print(arr)
quicksort_no_tr(arr, 0, n - 1, print_flag=True)
print(arr)

[6, 3, 2, 5, 7, 1, 0, 4]
[3, 2, 1, 0, 4, 6, 5, 7]
[0, 2, 1, 3, 4, 6, 5, 7]
[0, 2, 1, 3, 4, 6, 5, 7]
[0, 1, 2, 3, 4, 6, 5, 7]
[0, 1, 2, 3, 4, 6, 5, 7]
[0, 1, 2, 3, 4, 5, 6, 7]
[0, 1, 2, 3, 4, 5, 6, 7]


In [7]:
#
# In class 10/1/25 - time Quicksort (without tail recursion) on random data for a variety of array sizes.
#

rng = np.random.default_rng()
TEST_RUNS = 10
sizes = [10, 50, 100, 1000, 2500, 7500, 10000, 12500, 17500, 20000]

for i in range(TEST_RUNS):
        n = sizes[i]
        arr = rng.integers(low=0, high=100, size=n).tolist()

        start = time.perf_counter()
        quicksort_no_tr(arr, 0, n - 1)
        end = time.perf_counter()

        print(f"n = {n:<8,} time = {(end - start) * 1_000:.3f} ms")

n = 10       time = 0.016 ms
n = 50       time = 0.050 ms
n = 100      time = 0.095 ms
n = 1,000    time = 1.481 ms
n = 2,500    time = 7.496 ms
n = 7,500    time = 40.384 ms
n = 10,000   time = 65.748 ms
n = 12,500   time = 97.573 ms
n = 17,500   time = 192.487 ms
n = 20,000   time = 257.742 ms


| $n$    | $T(n)$      | $\dfrac{T(n)}{n\lg n}$      |
| ------ | ----------: | --------------------------: |
| 10     |       0.016 |                  0.00048165 |
| 50     |       0.050 |                  0.00017718 |
| 100    |       0.095 |                  0.00014299 |
| 1,000  |       1.481 |                  0.00014861 |
| 2,500  |       7.496 |                  0.00026563 |
| 7,500  |      40.384 |                  0.00041829 |
| 10,000 |      65.748 |                  0.00049480 |
| 12,500 |      97.573 |                  0.00057355 |
| 17,500 |     192.487 |                  0.00078036 |
| 20,000 |     257.742 |                  0.00090197 |


In [9]:
# Partition with a median of three approach

def median_of_three(arr: list[Any], p: int, r: int) -> int:
    mid = (p + r) // 2
    if (arr[p] > arr[mid]) ^ (arr[p] > arr[r]):
        return p
    elif (arr[mid] < arr[p]) ^ (arr[mid] < arr[r]):
        return mid
    else:
        return r

def partition_median(arr: list[Any], p: int, r: int) -> int:
    # Use median of three to select pivot and swap it with the last element
    median_index = median_of_three(arr, p, r)
    arr[median_index], arr[r] = arr[r], arr[median_index]

    x = arr[r]
    i = p - 1

    for j in range(p, r):
        if arr[j] <= x:
            i += 1
            arr[i], arr[j] = arr[j], arr[i]    # Swap arr[i] and arr[j]

    arr[i+1], arr[r] = arr[r], arr[i+1]        # Swap arr[i+1] and arr[r]

    return i + 1

In [10]:
# Implement Quicksort without tail recusion and with a median of three approach
def quicksort_mo3(arr: list[Any], p: int, r: int, *, print_flag: bool = False) -> None:
    while p < r:
      q = partition_median(arr, p, r)

      if print_flag:
        print(arr)

      quicksort_mo3(arr, p, q - 1, print_flag=print_flag)
      p = q + 1

n = 8
arr = list(range(n))
random.shuffle(arr)

print(arr)
quicksort_mo3(arr, 0, n - 1, print_flag=True)
print(arr)

[3, 7, 2, 6, 1, 5, 0, 4]
[3, 2, 1, 0, 4, 5, 6, 7]
[0, 1, 2, 3, 4, 5, 6, 7]
[0, 1, 2, 3, 4, 5, 6, 7]
[0, 1, 2, 3, 4, 5, 6, 7]
[0, 1, 2, 3, 4, 5, 6, 7]
